File: Problems/No.2556_Disconnect_Path_in_a_Binary_Matrix_by_at_Most_One_Flip/disconnect_path_in_a_binary_matrix_by_at_most_one_flip.cpp
class Solution {
public:

    // Depth-First Search
    bool DFS(std::vector<std::vector<int>> &grid, int row, int col, int m, int n) {

        if ((row >= m) || (col >= n) || (grid[row][col] == 0)) return false;
        else if ((row == m-1) && (col == n-1)) return true;
        else {

            grid[row][col] = 0;
            return (DFS(grid, row+1, col, m, n) || DFS(grid, row, col+1, m, n));

        }

    }

    bool isPossibleToCutPath(vector<vector<int>>& grid) {
        
        // parameters
        int paths = 0;
        int m = grid.size(), n = grid[0].size();

        // start computing
        if (grid[0][0] == 0) return true;
        paths = DFS(grid, 0, 0, m, n);
        grid[0][0] = 1;
        paths += DFS(grid, 0, 0, m, n);

        return (paths < 2);

    }
};

File: Problems/No.2556_Disconnect_Path_in_a_Binary_Matrix_by_at_Most_One_Flip/disconnect_path_in_a_binary_matrix_by_at_most_one_flip.py
class Solution(object):
    def isPossibleToCutPath(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: bool
        """

        # Depth-First Search
        def DFS(row, col):
            if ((row >= m) or (col >= n) or (grid[row][col] == 0)):
                return False
            elif ((row == m-1) and (col == n-1)):
                return True
            else:
                grid[row][col] = 0
                return (DFS(row+1, col) or DFS(row, col+1))
        

        # parameters
        paths = 0
        m, n = len(grid), len(grid[0])

        # start computing
        if (grid[0][0] == 0):
            return True
        paths = DFS(0, 0)
        grid[0][0] = 1
        paths += DFS(0, 0)

        return (paths < 2)
